File: sdks/typescript/packages/sdk/test/handler.test.ts
import { describe, it, expect } from 'vitest'
import { handleRequest } from '../src/handler.js'
import { signBody } from '../src/hmac.js'
import { signRefs } from '../src/refs.js'
import { defineScenario } from '../src/scenario.js'
import type { HandlerConfig } from '../src/types.js'

function createConfig(overrides?: Partial<HandlerConfig>): HandlerConfig {
  return {
    sharedSecret: 'test-secret',
    signingSecret: 'test-signing-secret',
    ...overrides,
  }
}

function signedRequest(body: Record<string, unknown>, secret: string) {
  const raw = JSON.stringify(body)
  return {
    body: raw,
    headers: { 'x-signature': signBody(raw, secret) },
  }
}

const singleUser = defineScenario({
  name: 'single-user',
  description: 'One user in a fresh org',
  up: async ({ testRunId }) => ({
    auth: { headers: { Authorization: `Bearer jwt-${testRunId}` } },
    teardown: { userId: `user-${testRunId}` },
  }),
})

describe('handleRequest (v2)', () => {
  describe('HMAC', () => {
    it('rejects invalid signature', async () => {
      const config = createConfig()
      const res = await handleRequest(config, {
        body: '{"action":"discover"}',
        headers: { 'x-signature': 'bad' },
      })
      expect(res.status).toBe(401)
      expect(res.body.code).toBe('INVALID_SIGNATURE')
    })

    it('rejects missing signature', async () => {
      const config = createConfig()
      const res = await handleRequest(config, {
        body: '{"action":"discover"}',
        headers: {},
      })
      expect(res.status).toBe(401)
    })

    it('rejects identical shared and signing secrets', async () => {
      const config = createConfig({ sharedSecret: 'same', signingSecret: 'same' })
      const req = signedRequest({ action: 'discover' }, 'same')
      const res = await handleRequest(config, req)
      expect(res.status).toBe(500)
      expect(res.body.code).toBe('SAME_SECRETS')
    })
  })

  describe('discover', () => {
    it('returns an empty scenario list when none registered', async () => {
      const config = createConfig()
      const req = signedRequest({ action: 'discover' }, config.sharedSecret)
      const res = await handleRequest(config, req)

      expect(res.status).toBe(200)
      const body = res.body as Record<string, any>
      expect(body.scenarios).toEqual([])
      expect(body.version).toBe('2.0')
    })

    it('lists registered scenarios as { name, description }', async () => {
      const config = createConfig({
        scenarios: [
          singleUser,
          defineScenario({ name: 'empty', description: 'Nothing seeded', up: () => ({}) }),
        ],
      })
      const req = signedRequest({ action: 'discover' }, config.sharedSecret)
      const res = await handleRequest(config, req)

      expect(res.status).toBe(200)
      const body = res.body as Record<string, any>
      expect(body.scenarios).toEqual([
        { name: 'single-user', description: 'One user in a fresh org' },
        { name: 'empty', description: 'Nothing seeded' },
      ])
      // The factory-derived model schema is gone.
      expect(body.schema).toBeUndefined()
    })
  })

  describe('up', () => {
    it('runs the scenario up and returns auth, teardownToken, expiry', async () => {
      const config = createConfig({ scenarios: [singleUser] })
      const req = signedRequest(
        { action: 'up', scenario: { name: 'single-user' }, testRunId: 'run-123' },
        config.sharedSecret,
      )
      const res = await handleRequest(config, req)

      expect(res.status).toBe(200)
      const body = res.body as Record<string, any>
      expect(body.version).toBe('2.0')
      expect(body.auth.headers.Authorization).toBe('Bearer jwt-run-123')
      expect(typeof body.teardownToken).toBe('string')
      // The duplicated plaintext refs and the old refsToken field are gone.
      expect(body.refs).toBeUndefined()
      expect(body.refsToken).toBeUndefined()
      expect(body.expiresInSeconds).toBe(3600)
    })

    it('applies a configured default expiry', async () => {
      const config = createConfig({ scenarios: [singleUser], expiresInSeconds: 900 })
      const req = signedRequest(
        { action: 'up', scenario: { name: 'single-user' }, testRunId: 'r' },
        config.sharedSecret,
      )
      const res = await handleRequest(config, req)
      expect((res.body as Record<string, any>).expiresInSeconds).toBe(900)
    })

    it('omits auth when the scenario returns none', async () => {
      const config = createConfig({
        scenarios: [defineScenario({ name: 'bare', description: 'x', up: () => ({}) })],
      })
      const req = signedRequest(
        { action: 'up', scenario: { name: 'bare' }, testRunId: 'r' },
        config.sharedSecret,
      )
      const res = await handleRequest(config, req)
      expect(res.status).toBe(200)
      const body = res.body as Record<string, any>
      expect(body.auth).toBeUndefined()
      expect(typeof body.teardownToken).toBe('string')
    })

    it('throws UNKNOWN_ENVIRONMENT for an unregistered scenario name', async () => {
      const config = createConfig({ scenarios: [singleUser] })
      const req = signedRequest(
        { action: 'up', scenario: { name: 'does-not-exist' }, testRunId: 'r' },
        config.sharedSecret,
      )
      const res = await handleRequest(config, req)
      expect(res.status).toBe(400)
      expect(res.body.code).toBe('UNKNOWN_ENVIRONMENT')
    })

    it('returns 400 when scenario.name is missing', async () => {
      const config = createConfig({ scenarios: [singleUser] })
      const req = signedRequest({ action: 'up', testRunId: 'r' }, config.sharedSecret)
      const res = await handleRequest(config, req)
      expect(res.status).toBe(400)
      expect(res.body.code).toBe('INVALID_BODY')
    })
  })

  describe('down', () => {
    it('routes to the scenario down with the teardown handle from the token', async () => {
      const captured: { name?: string; teardown?: unknown; testRunId?: string } = {}
      const scenario = defineScenario({
        name: 'teardownable',
        description: 'x',
        up: async ({ testRunId }) => ({ teardown: { handle: `h-${testRunId}` } }),
        down: async ({ name, teardown, testRunId }) => {
          captured.name = name
          captured.teardown = teardown
          captured.testRunId = testRunId
        },
      })
      const config = createConfig({ scenarios: [scenario] })

      const upRes = await handleRequest(
        config,
        signedRequest(
          { action: 'up', scenario: { name: 'teardownable' }, testRunId: 'run-x' },
          config.sharedSecret,
        ),
      )
      const teardownToken = (upRes.body as Record<string, any>).teardownToken

      const downRes = await handleRequest(
        config,
        signedRequest(
          { action: 'down', teardownToken, testRunId: 'run-x' },
          config.sharedSecret,
        ),
      )
      expect(downRes.status).toBe(200)
      expect((downRes.body as Record<string, any>).ok).toBe(true)
      expect(captured.name).toBe('teardownable')
      expect(captured.teardown).toEqual({ handle: 'h-run-x' })
      expect(captured.testRunId).toBe('run-x')
    })

    it('recovers the scenario name from the token when the request omits it', async () => {
      let downRan = false
      const scenario = defineScenario({
        name: 'from-token',
        description: 'x',
        up: () => ({ teardown: {} }),
        down: () => {
          downRan = true
        },
      })
      const config = createConfig({ scenarios: [scenario] })
      const teardownToken = signRefs(
        { refs: {}, testRunId: 'r', environment: 'from-token' },
        config.signingSecret,
      )
      const res = await handleRequest(
        config,
        signedRequest({ action: 'down', teardownToken }, config.sharedSecret),
      )
      expect(res.status).toBe(200)
      expect(downRan).toBe(true)
    })

    it('no-ops when the scenario defines no down', async () => {
      const scenario = defineScenario({
        name: 'no-down',
        description: 'x',
        up: () => ({ teardown: {} }),
      })
      const config = createConfig({ scenarios: [scenario] })
      const teardownToken = signRefs(
        { refs: {}, testRunId: 'r', environment: 'no-down' },
        config.signingSecret,
      )
      const res = await handleRequest(
        config,
        signedRequest({ action: 'down', teardownToken }, config.sharedSecret),
      )
      expect(res.status).toBe(200)
      expect((res.body as Record<string, any>).ok).toBe(true)
    })

    it('rejects a tampered teardownToken', async () => {
      const config = createConfig()
      const req = signedRequest(
        { action: 'down', teardownToken: 'bad.token.here' },
        config.sharedSecret,
      )
      const res = await handleRequest(config, req)
      expect(res.status).toBe(403)
      expect(res.body.code).toBe('INVALID_TEARDOWN_TOKEN')
    })
  })

  describe('errors', () => {
    it('returns 400 for unknown action', async () => {
      const config = createConfig()
      const req = signedRequest({ action: 'invalid' }, config.sharedSecret)
      const res = await handleRequest(config, req)
      expect(res.status).toBe(400)
      expect(res.body.code).toBe('UNKNOWN_ACTION')
    })

    it('returns 400 for invalid JSON', async () => {
      const config = createConfig()
      const body = 'not json'
      const res = await handleRequest(config, {
        body,
        headers: { 'x-signature': signBody(body, config.sharedSecret) },
      })
      expect(res.status).toBe(400)
      expect(res.body.code).toBe('INVALID_BODY')
    })
  })
})

File: sdks/elixir/test/autonoma/handler_test.exs
defmodule Autonoma.HandlerTest do
  use ExUnit.Case, async: true

  alias Autonoma.{Handler, HMAC, Scenario}

  @shared_secret "test-shared-secret-1234"
  @signing_secret "test-signing-secret-5678"

  defp standard_scenario(down \\ nil) do
    Scenario.define_scenario(
      name: "standard",
      description: "A standard seeded environment",
      up: fn ctx ->
        %{
          auth: %{"headers" => %{"Authorization" => "Bearer #{ctx.test_run_id}"}},
          teardown: %{"userId" => "user-#{ctx.test_run_id}"}
        }
      end,
      down: down
    )
  end

  defp empty_scenario do
    Scenario.define_scenario(
      name: "empty",
      description: "Nothing seeded",
      up: fn _ctx -> %{} end
    )
  end

  defp make_config(overrides \\ %{}) do
    Map.merge(
      %{
        shared_secret: @shared_secret,
        signing_secret: @signing_secret,
        scenarios: [standard_scenario(), empty_scenario()]
      },
      overrides
    )
  end

  defp signed_req(body_map) do
    body = Jason.encode!(body_map)
    %{body: body, headers: %{"x-signature" => HMAC.sign_body(body, @shared_secret)}}
  end

  # --- request gate ---

  test "rejects invalid signature" do
    req = %{body: ~s({"action":"discover"}), headers: %{"x-signature" => "invalid"}}
    result = Handler.handle(make_config(), req)
    assert result.status == 401
    assert result.body["code"] == "INVALID_SIGNATURE"
  end

  test "rejects same secrets" do
    config = %{shared_secret: "same", signing_secret: "same", scenarios: []}
    body = ~s({"action":"discover"})
    req = %{body: body, headers: %{"x-signature" => HMAC.sign_body(body, "same")}}
    result = Handler.handle(config, req)
    assert result.status == 500
    assert result.body["code"] == "SAME_SECRETS"
  end

  test "rejects invalid JSON" do
    raw = "not json"
    req = %{body: raw, headers: %{"x-signature" => HMAC.sign_body(raw, @shared_secret)}}
    result = Handler.handle(make_config(), req)
    assert result.status == 400
    assert result.body["code"] == "INVALID_BODY"
  end

  test "rejects missing action" do
    result = Handler.handle(make_config(), signed_req(%{"foo" => "bar"}))
    assert result.status == 400
    assert result.body["code"] == "INVALID_BODY"
  end

  test "rejects unknown action" do
    result = Handler.handle(make_config(), signed_req(%{"action" => "nonexistent"}))
    assert result.status == 400
    assert result.body["code"] == "UNKNOWN_ACTION"
  end

  # --- discover ---

  test "discover lists registered scenarios" do
    result = Handler.handle(make_config(), signed_req(%{"action" => "discover"}))
    assert result.status == 200
    assert result.body["version"] == "2.0"
    assert result.body["sdk"]["language"] == "elixir"
    assert length(result.body["scenarios"]) == 2
    assert List.first(result.body["scenarios"])["name"] == "standard"
    assert List.first(result.body["scenarios"])["description"] != ""
    # discover must never leak a create/schema shape in v2.
    refute Map.has_key?(result.body, "schema")
  end

  # --- up ---

  test "up returns the envelope" do
    result =
      Handler.handle(
        make_config(),
        signed_req(%{"action" => "up", "scenario" => %{"name" => "standard"}, "testRunId" => "run-1"})
      )

    assert result.status == 200
    assert result.body["version"] == "2.0"
    assert length(String.split(result.body["teardownToken"], ".")) == 3
    assert result.body["expiresInSeconds"] == 3600
    # The duplicated plaintext refs and the old refsToken field are gone.
    refute Map.has_key?(result.body, "refs")
    refute Map.has_key?(result.body, "refsToken")
    assert result.body["auth"]["headers"]["Authorization"] == "Bearer run-1"
  end

  test "up honors a custom expiresInSeconds and omits auth when empty" do
    config = make_config(%{expires_in_seconds: 60})

    result =
      Handler.handle(
        config,
        signed_req(%{"action" => "up", "scenario" => %{"name" => "empty"}, "testRunId" => "r"})
      )

    assert result.status == 200
    assert result.body["expiresInSeconds"] == 60
    refute Map.has_key?(result.body, "auth")
  end

  test "up rejects an unknown environment" do
    result =
      Handler.handle(
        make_config(),
        signed_req(%{"action" => "up", "scenario" => %{"name" => "does-not-exist"}, "testRunId" => "r"})
      )

    assert result.status == 400
    assert result.body["code"] == "UNKNOWN_ENVIRONMENT"
  end

  test "up rejects a missing scenario name" do
    result = Handler.handle(make_config(), signed_req(%{"action" => "up", "testRunId" => "r"}))
    assert result.status == 400
    assert result.body["code"] == "INVALID_BODY"
  end

  test "up accepts a string-keyed result map" do
    string_keyed =
      Scenario.define_scenario(
        name: "string-keyed",
        description: "up returns a string-keyed map",
        up: fn _ctx ->
          %{"auth" => %{"headers" => %{"X-Token" => "plain"}}}
        end
      )

    config = make_config(%{scenarios: [string_keyed]})

    result =
      Handler.handle(
        config,
        signed_req(%{"action" => "up", "scenario" => %{"name" => "string-keyed"}, "testRunId" => "r"})
      )

    assert result.status == 200
    assert result.body["auth"]["headers"]["X-Token"] == "plain"
  end

  test "up defaults testRunId when absent" do
    result =
      Handler.handle(
        make_config(),
        signed_req(%{"action" => "up", "scenario" => %{"name" => "empty"}})
      )

    assert result.status == 200
    assert is_binary(result.body["teardownToken"])
  end

  # --- down ---

  test "down tears a scenario down with a valid teardownToken" do
    test_pid = self()
    down = fn ctx -> send(test_pid, {:down, ctx.name, ctx.test_run_id}) end
    config = make_config(%{scenarios: [standard_scenario(down), empty_scenario()]})

    up =
      Handler.handle(
        config,
        signed_req(%{"action" => "up", "scenario" => %{"name" => "standard"}, "testRunId" => "run-td"})
      )

    token = up.body["teardownToken"]

    down_result =
      Handler.handle(
        config,
        signed_req(%{"action" => "down", "teardownToken" => token, "testRunId" => "run-td"})
      )

    assert down_result.status == 200
    assert down_result.body["ok"] == true
    assert_receive {:down, "standard", "run-td"}
  end

  test "down routes by the token environment, ignoring any request scenario name" do
    test_pid = self()
    down = fn ctx -> send(test_pid, {:down, ctx.name, ctx.test_run_id}) end
    config = make_config(%{scenarios: [standard_scenario(down), empty_scenario()]})

    up =
      Handler.handle(
        config,
        signed_req(%{"action" => "up", "scenario" => %{"name" => "standard"}, "testRunId" => "run-tok"})
      )

    token = up.body["teardownToken"]

    # No scenario.name on the down request - the handler must recover it from
    # the verified token's environment.
    down_result =
      Handler.handle(config, signed_req(%{"action" => "down", "teardownToken" => token}))

    assert down_result.status == 200
    assert_receive {:down, "standard", "run-tok"}
  end

  test "down rejects an invalid teardownToken" do
    result =
      Handler.handle(
        make_config(),
        signed_req(%{"action" => "down", "teardownToken" => "tampered.token.value"})
      )

    assert result.status == 403
    assert result.body["code"] == "INVALID_TEARDOWN_TOKEN"
  end

  test "down rejects a missing teardownToken" do
    result = Handler.handle(make_config(), signed_req(%{"action" => "down"}))
    assert result.status == 400
    assert result.body["code"] == "INVALID_BODY"
  end

  # --- deprecated allow_production ---

  test "endpoint is always enabled and ignores the deprecated allow_production flag" do
    config = make_config(%{allow_production: false})
    result = Handler.handle(config, signed_req(%{"action" => "discover"}))
    assert result.status == 200
    assert length(result.body["scenarios"]) == 2
  end
end
